src/test/spec/unified_runner/mod.rs

```diff
@@ -36,6 +36,109 @@ lazy_static! {
     static ref SPEC_VERSIONS: Vec<Version> = vec![Version::parse("1.0.0").unwrap()];
 }
 
+pub struct TestRunner {
+    pub internal_client: TestClient,
+    pub entities: HashMap<String, Entity>,
+    pub failpoint_disable_commands: Vec<Document>,
+}
+
+impl TestRunner {
+    pub async fn new() -> Self {
+        Self {
+            internal_client: TestClient::new().await,
+            entities: HashMap::new(),
+            failpoint_disable_commands: Vec::new(),
+        }
+    }
+
+    pub async fn insert_initial_data(&self, data: &CollectionData) {
+        let write_concern = WriteConcern::builder().w(Acknowledgment::Majority).build();
+        let collection_options = CollectionOptions::builder()
+            .write_concern(write_concern.clone())
+            .build();
+        let coll = self
+            .internal_client
+            .init_db_and_coll_with_options(
+                &data.database_name,
+                &data.collection_name,
+                collection_options,
+            )
+            .await;
+
+        let insert_options = InsertManyOptions::builder()
+            .write_concern(write_concern)
+            .build();
+        if !data.documents.is_empty() {
+            coll.insert_many(data.documents.clone(), insert_options)
+                .await
+                .unwrap();
+        }
+    }
+
+    pub async fn disable_fail_points(&self) {
+        for disable_command in &self.failpoint_disable_commands {
+            self.internal_client
+                .database("admin")
+                .run_command(disable_command.clone(), None)
+                .await
+                .unwrap();
+        }
+    }
+
+    pub async fn populate_entity_map(&mut self, create_entities: &[TestFileEntity]) {
+        for entity in create_entities {
```

Comment:
I don't think this will really be a problem, but we probably should clear out the entity map before populating it to ensure each run has completely new state.